source/core/lethegridtools.cc

```diff
@@ -0,0 +1,982 @@
+                        best_dist  = dist;
+                        best_index = i;
+                        cell_found = true;
+                        if (dist == 0)
+                            break;
+                    }
+                }
+                catch (
+                        const typename MappingQGeneric<dim>::ExcTransformationFailed &)
+                {}
+            }
+
+            best_cell_iter = best_cell_iter->child(best_index);
+
+            if (cell_found == false)
+            {
+                std::cout << "Cell not found around " << point << std::endl;
+                break;
+            }
+
+            lvl += 1;
+        }
+    }
+
+
+    return best_cell_iter;
+}
+
+template <int dim>
+typename DoFHandler<dim>::active_cell_iterator
+LetheGridTools::find_cell_around_point_with_neighbors(const DoFHandler<dim> &dof_handler,
+                                                      std::map<unsigned int,std::set<typename DoFHandler<dim>::active_cell_iterator>> &vertices_cell_map,
+        const typename DoFHandler<dim>::active_cell_iterator &cell,
+        Point<dim>                                            &point)
+{
+    // Find the cell around a point based on an initial cell.
+
+    // Find the cells around the initial cell ( cells that share a vertex with the
+    // original cell).
+    std::vector<typename DoFHandler<dim>::active_cell_iterator>
+            active_neighbors_set = LetheGridTools::find_cells_around_cell<dim>(vertices_cell_map,cell);
+    // Loop over that group of cells
+    for (unsigned int i = 0; i < active_neighbors_set.size(); ++i)
+    {
+        bool inside_cell = active_neighbors_set[i]->point_inside(point);
+        if (inside_cell)
+        {
+            return active_neighbors_set[i];
+        }
+    }
+    // The cell is not found near the initial cell so we use the cell tree
+    // algorithm instead (much slower).
+    std::cout << "Cell not found around " << point << std::endl;
+    return LetheGridTools::find_cell_around_point_with_tree(dof_handler, point);
+}
+
+template <int dim>
+std::vector<typename DoFHandler<dim>::active_cell_iterator>
+LetheGridTools::find_cells_around_cell(std::map<unsigned int,std::set<typename DoFHandler<dim>::active_cell_iterator>> &vertices_cell_map,
+                                                        const typename DoFHandler<dim>::active_cell_iterator &cell)
+{
+    // Find all the cells that share a vertex with a reference cell including the
+    // initial cell.
+    //std::unordered_set<typename DoFHandler<dim>::active_cell_iterator, LetheGridTools::hash_cell<dim>, LetheGridTools::equal_cell<dim>> neighbors_cells;
+    std::set<typename DoFHandler<dim>::active_cell_iterator> neighbors_cells;
+
+    // Loop over the vertices of the initial cell and find all the cells around
+    // each vertex and add them to the set of cells around the reference cell.
+    for (unsigned int i = 0; i < GeometryInfo<dim>::vertices_per_cell; i++)
```

Comment:
Be careful, this only works for the case of Hexahedral triangulation. It's not that big of a deal but it could eventually be. https://www.dealii.org/developer/doxygen/deal.II/group__simplex.html#details

Reply:
Good point I suggest we add this option in the next PR as well.